---
title: "R Notebook"
output:
  html_document:
    df_print: paged
---
#BACTERIA DIVERSITY AND COMPOSITION
##Activating libraries
```{r}
library(tidyverse)
library("ggsci")
library(ggpubr)
library(rstatix)
library(scales)
```

#Importing tables
```{r}
setwd("~/OneDrive\ -\ Michigan\ State\ University/Manning_lab/Metabolomics_mastitis_microbial/tables/biobakery3/")

library(readxl)
#read in otu table (smartchip normalized abundance)
otu_table <- read_excel("phyloseq_tables_biobakery.xlsx", sheet = "humann_pathabundance_relab")

library(dplyr)
library(tibble)
#converting the abundance table to matrix (required to merge with taxonomy table)
otu_table <- otu_table %>% remove_rownames %>% column_to_rownames(var="OTU_ID") 
otu_table=as.matrix(otu_table,rownames=TRUE)
otu_table <- otu_table*10^9
#otu_table[otu_table >0]<-1
mode(otu_table) <- "integer"

#converting empty values to 0 (in case they exist)
otu_table[is.na(otu_table)] <- 0

#read in taxonomy (gene ID for smartchip)
taxonomy <- read_excel("phyloseq_tables_biobakery.xlsx", sheet = "taxonomy_path")

#setting taxonomy row_names for GENE_ID (that will enable merging with the abundance table)
taxonomy <- taxonomy %>% remove_rownames %>% column_to_rownames(var="OTU_ID")
taxonomy=as.matrix(taxonomy)

#read in metadata
metadata <- read_excel("phyloseq_tables_biobakery.xlsx", sheet = "metadata")

#setting metadata row_names for sample ID (that will enable merging with the other tables)
metadata <- metadata %>% remove_rownames %>% column_to_rownames(var="sequence_id")
metadata <- metadata %>% mutate(Treatment = fct_relevel(Treatment,"Control","Antibiotic"),
                                Time_tx = fct_relevel(Time_tx,"Day -1","Week 1","Week 5","Week 9"),
                                Group_Sampling = fct_relevel(Group_Sampling, "Control Day -1", "Antibiotic Day -1", "Control Week 1", "Antibiotic Week 1","Control Week 5", "Antibiotic Week 5","Control Week 9", "Antibiotic Week 9"))


library(forcats)
#Changing the levels of source and group to make them better organized in the plots
metadata <- metadata %>% mutate(Treatment = fct_relevel(Treatment,"Control","Antibiotic"))

library("phyloseq")

#import all tables as phyloseq objects
OTU = otu_table(otu_table,taxa_are_rows=TRUE)
TAX = tax_table(taxonomy)
META = sample_data(metadata)

# Sanity checks for consistent OTU names
taxa_names(TAX)
taxa_names(OTU)

# Same sample names
sample_names(OTU)
sample_names(META)

#Making phyloseq object
physeq=phyloseq(OTU,TAX,META)
```

```{r}
`%notin%` <- Negate(`%in%`)

#Filtering samples
physeq_filtered = subset_taxa(physeq, Pathway %notin% "UNMAPPED")
physeq = physeq_filtered
#physeq = subset_samples(physeq_hindgut, Farm%in%c("8B","9D"))
```

#Alpha diversity
```{r fig.width=10, fig.height=40}
alpha_diversity <- estimate_richness(physeq, measures = c("Shannon", "Observed"))
df_alpha <- data.frame(alpha_diversity, sample_data(physeq))
df <- reshape2::melt(df_alpha, measure.var=c("Shannon","Observed"),id.vars=c("Treatment","Time_tx", "study_ID","sample_ID"))
df$value = as.numeric(df$value)

alpha <- df %>% rename(Index = variable)
#selecting the rows corresponding to each index from the main dataframe
shannon = dplyr::filter(df, variable == "Shannon")
Observed = dplyr::filter(df, variable == "Observed")

```
```{r}
library(rstatix)
alpha %>% 
  group_by(Index) %>% 
  shapiro_test(value)
```
```{r}
library(rstatix)
alpha %>% 
  group_by(Index, Time_tx) %>% 
  shapiro_test(value)
```

#SHANNON
##Significance over the time period 
All
```{r}
stat.test <- shannon %>%
   filter(study_ID %notin% "MA029") %>% 
  anova_test(value ~ Time_tx) 
stat.test
```

per group
```{r}
stat.test <- shannon %>%
   filter(study_ID %notin% "MA029") %>% 
  group_by(Treatment) %>% 
  anova_test(value ~ Time_tx) 
stat.test
```

```{r}
#Calculating p-values between treatments by time point
stat.test <- shannon %>%
  filter(sample_ID %notin% "MA028.7") %>% 
  group_by(Time_tx) %>%
  t_test(value ~ Treatment, alternative = "greater", paired = T) %>%
  adjust_pvalue(method = "none") %>%
  add_significance("p.adj") %>% 
  add_xy_position(x = "Time_tx", dodge = 0.8)
stat.test
```

```{r}
#Calculating p-values between treatments by time point
stat.test <- Observed %>%
  filter(sample_ID %notin% "MA028.7") %>% 
  group_by(Time_tx) %>%
  t_test(value ~ Treatment, alternative = "greater", paired = T) %>%
  adjust_pvalue(method = "none") %>%
  add_significance("p.adj") %>% 
  add_xy_position(x = "Time_tx", dodge = 0.8)
stat.test
```

```{r}
#Calculating p-values between treatments by time point
stat.test <- shannon %>%
  filter(sample_ID %notin% "MA028.7") %>% 
  group_by(Time_tx) %>%
  t_test(value ~ Treatment, alternative = "greater", paired = T) %>%
  adjust_pvalue(method = "none") %>%
  add_significance("p.adj") %>% 
  add_xy_position(x = "Time_tx", dodge = 0.8)
stat.test
```

#Alpha diversity gut plot
```{r fig.width=5, fig.height=5}
library(ggsci)

my_comparisons <- list(c("Day -1", "Week 1"), c("Week 1", "Week 5"), c("Week 5", "Week 9"), c("Day -1", "Week 9"))

##################### Running box plot #############################

shannon_path <- shannon %>% 
 ggboxplot(x = "Time_tx", y = "value", color = "Treatment", palette = "aaas", 
            fill = "Treatment", add = c("jitter"), notch = F, outlier.shape = NA, alpha = 0.5,facet.by = "variable", panel.labs = list(variable=c("Microbial Pathways","Microbial Pathways"))) +
  labs(x = "Time to treatment", y = "Shannon index") + 
  theme(legend.position="top") + 
  stat_pvalue_manual(stat.test,  label = "p", tip.length = 0) +
  annotate("text", x = 2.5, y = 2.6,  label = expression(paste("ANOVA, ", paste(italic('p'))," = 1.09e-09"))) +
#   annotate("text", x = 1.5, y = 2.6, label = expression(paste("Control (FT), ", paste(italic('p'))," = 0.004")),            color = "#374e55", size = 3) +
 # annotate("text", x = 3.5, y = 2.6, label = expression(paste("Antibiotic (FT), ", paste(italic('p'))," = 0.0001")),            color = "#df8f44", size = 3) + 
  stat_compare_means(comparisons = my_comparisons, label.y = c(2.2,2,2.2, 2.35), method = "t.test", paired = F,
                     tip.length=.02, hide.ns = T, color = "gray35") +
    theme(axis.title.x = element_blank())

shannon_path
```

#Observed
##Significance over the time period 
All
```{r}
stat.test <- Observed %>%
   filter(study_ID %notin% "MA029") %>% 
  anova_test(value ~ Time_tx) 
stat.test
```

per group
```{r}
stat.test <- Observed %>%
   filter(study_ID %notin% "MA029") %>% 
  group_by(Treatment) %>% 
  anova_test(value ~ Time_tx) 
stat.test
```

```{r}
#Calculating p-values between treatments by time point
stat.test <- Observed %>%
  filter(sample_ID %notin% "MA028.7") %>% 
  group_by(Time_tx) %>%
  t_test(value ~ Treatment, alternative = "greater", paired = T) %>%
  adjust_pvalue(method = "none") %>%
  add_significance("p.adj") %>% 
  add_xy_position(x = "Time_tx", dodge = 0.8)

stat.test
```

#Alpha diversity gut plot
```{r fig.width=5, fig.height=5}
library(ggsci)

my_comparisons <- list(c("Day -1", "Week 1"), c("Week 1", "Week 5"), c("Week 5", "Week 9"), c("Day -1", "Week 9"))

##################### Running box plot #############################

observed_path <- Observed %>% 
 ggboxplot(x = "Time_tx", y = "value", color = "Treatment", palette = "aaas", 
            fill = "Treatment", add = c("jitter"), notch = F, outlier.shape = NA, alpha = 0.5,facet.by = "variable", panel.labs = list(variable=c("Microbial Pathways","Microbial Pathways"))) +
  labs(x = "Time to treatment", y = "Observed") + 
  theme(legend.position="top") + 
  stat_pvalue_manual(stat.test,  label = "p", tip.length = 0) +
  annotate("text", x = 2.5, y = 630, label = expression(paste("ANOVA, ", paste(italic('p'))," = 3.36e-18"))) +
#   annotate("text", x = 1.5, y = 590, label = expression(paste("Control (FT), ", paste(italic('p'))," = 7.22e-04")),            color = "#374e55", size = 3) +
#  annotate("text", x = 3.5, y = 590, label = expression(paste("Antibiotic (FT), ", paste(italic('p'))," = 1.12e-05")),            color = "#df8f44", size = 3) + 
  stat_compare_means(comparisons = my_comparisons, label.y = c(500,300,450,550), 
                     tip.length=.02, hide.ns = T, color = "gray35") +   
  theme(axis.title.x = element_blank())

observed_path
```

```{r}
alpha_path <- ggarrange(shannon_plot, Observed_plot, labels = c("A","B"), nrow=1, common.legend = T)
alpha_path
```

```{r}
setwd("~/OneDrive\ -\ Michigan\ State\ University/Manning_lab/Metabolomics_mastitis/figures/biobakery")
ggsave(plot = alpha_path, "alpha_diversity_pathways.png",width=10,height=6)
```

##Beta-diversity

##Bray-Curtis plot
```{r}
beta <-betatest(physeq,group="Group_Sampling")
beta
#Comparison_clusters
```
```{r}
beta <-betatest(physeq,group="Comparison_clusters")
beta
#Comparison_clusters
```
```{r}
beta <-betatest(physeq,group="Treatment")
beta
```

```{r}
bray_path<- plotbeta(
  physeq,
  group="Group_Sampling",
  shape = "Treatment",
  distance = "bray",
  method = "PCoA",
  color = NULL,
  size = 3,
  ellipse = TRUE) + 
  labs(color = "Group", shape = "Treatment") + 
  annotate("text", x = 0.1, y = 0.18, label = expression(paste("PERMANOVA, ",F ,"= 63.69, ",paste(italic('p')),"=0.001")), colour = "black", size = 4) + ggtitle("Microbial Pathways") +
  theme_bw()
bray_path
```
```{r}
physeq_week1 <- subset_samples(physeq, Time_tx%in%c("Week 1"))
physeq_day1 <- subset_samples(physeq, Time_tx%in%c("Day -1"))
physeq_week9 <- subset_samples(physeq, Time_tx%in%c("Week 9"))
physeq_week5 <- subset_samples(physeq, Time_tx%in%c("Week 5"))
```

```{r}
beta <-betatest(physeq_week1,group="Treatment")
beta
```

###Bray statistics
```{r}
## PERMANOVA
library(vegan)
bray_dist = phyloseq::distance(physeq, method="bray") #RUN this only once because it takes a lot of time
adonis_bray_cohort = adonis(bray_dist ~ sample_data(physeq)$Cohort)
adonis_bray_cohort

adonis_bray_time = adonis(bray_dist ~ sample_data(physeq)$Time_tx)
adonis_bray_time
## Significant PERMANOVA indicates that centroid (or spatial median) among groups is different and/or with-group dispersion among groups is different

## PERMDISP
bray_disp_cohort <-betadisper(bray_dist, sample_data(physeq)$Comparison_clusters, type=c("median"))
anova(bray_disp_cohort)
TukeyHSD(bray_disp_cohort)

bray_disp_time <-betadisper(bray_dist, sample_data(physeq)$Time_tx, type=c("median"))

anova(bray_disp_time)
TukeyHSD(bray_disp_time)


## If PERMANOVA and PERMDISP are both significant, you can use plotting to tell if PERMANOVA was significant based on centroid (or spatial median)
plot(bray_disp_cohort)

plot(bray_disp_time)
#?plot.betadisper

## Plot with 1 standard deviation ellipses around the group medians
## sample size issue here, but you get the idea
plot(bray_disp_cohort, label = F, hull = F, ellipse = T)
plot(bray_disp_time, label = F, hull = F, ellipse = T)

## Within-group dispersion that PERMDISP is testing
boxplot(bray_disp_cohort, las = 2, cex.lab=1.5)
boxplot(bray_disp_time, las = 2, cex.lab=1.5)
```
##Bray beta diversity with phyloseq
```{r}
beta_bray <- ordinate(physeq, "PCoA", "bray") #RUN this only ONCE because it takes a lot of time
```

Antibiotic use
```{r fig.width=8}
beta_bray_plot_time= plot_ordination(physeq, beta_bray, type="Time_tx", color="Time_tx") +
  stat_ellipse(type = "t", lineSource = 3) + 
  labs(color = "Time point") +
  theme_classic()+
  scale_fill_jama() +
  scale_color_jama() + 
  annotate("text", x = 0.03, y = 0.04, label = expression(paste("PERMANOVA, ",R^2 ,"= 0.33, ", paste(italic('p')),"=0.001"))) +
  theme(legend.position="top")
  #To add arrows https://neavemj.github.io/posts/coralMicrobiome +

beta_bray_plot_time
```

Antibiotic use
```{r fig.width=8}
library(ggsci)
beta_bray_plot_cohort= plot_ordination(physeq, beta_bray, type="Cohort", color="Cohort") +
  stat_ellipse(type = "t", lineSource = 3) + 
  labs(color = "Cohort") +
  theme_classic()+
  scale_fill_simpsons() +
  scale_color_simpsons() + 
  annotate("text", x = 0.03, y = 0.04, label = expression(paste("PERMANOVA, ",R^2 ,"= 0.36, ", paste(italic('p')),"=0.001"))) +
  theme(legend.position="top")
  #To add arrows https://neavemj.github.io/posts/coralMicrobiome +

beta_bray_plot_cohort
```

```{r}
beta_plots <- ggarrange(beta_bray_plot_time, beta_bray_plot_cohort, labels = c("A","B"), nrow = 1, ncol = 2)
beta_plots
```

```{r}
setwd("~/OneDrive\ -\ Michigan\ State\ University/Manning_lab/Metabolomics_mastitis/figures/biobakery")
ggsave(plot = beta_plots, "beta_bray_pathways.png", width = 16, height = 4)
```

#COMPOSITION
```{r}
glom_path <- tax_glom(physeq, taxrank = 'Pathway')
data_path <- psmelt(glom_path) # create dataframe from phyloseq object

library(dplyr)
data_path_rel = data_path %>% group_by(Sample) %>% mutate(relAbundBySample = Abundance / sum(Abundance)*100) #Converting counts to relative abundance %
data_path_rel = data_path_rel %>% mutate(Abundance = Abundance / 10^9) #Converting counts to normalized abundance since I multiplied the normalized abundance by 10^9 to calculate the Shannon indes
```

```{r fig.width=12}

path_abn <- data_path_rel %>% 
  filter(Pathway %notin% "UNINTEGRATED") %>% 
  group_by(Pathway) %>% 
  summarise(abn = sum(Abundance))

top20 <- top_n(path_abn, 20)

path_ab_bar <- data_path_rel %>% 
  filter(Pathway %in% top20$Pathway) %>% 
  ggbarplot(x= "Time_tx", y = "Abundance", color = "Pathway", fill = "Pathway", palette = get_palette("simpsons",20), ylab = "Relative abundance", legend = "right", order = c("Day -1","Week 1","Week 5","Week 9")) + 
  facet_wrap(~Treatment, scales = "free_x") +
  labs(colour = "Pathway", fill = "Pathway")  

path_ab_bar
```

```{r}
setwd("~/OneDrive\ -\ Michigan\ State\ University/Manning_lab/Metabolomics_mastitis/figures/biobakery")
ggsave(plot = path_ab_bar, "top20_pathway_relab_barplot.png", width = 16, height = 8)
```

```{r fig.width=12}
ARG_class_ab_group_facet <- data_class_rel %>% 
  mutate(across(Group_Source, factor, levels =c("Control_Meconium", "Sepsis_Meconium", "Control_Infant 7 days", "Sepsis_Infant 7 days", "Control_Mother", "Sepsis_Mother"))) %>%
  ggbarplot(x= "Sample", y = "relAbundBySample", color = "Class", fill = "Class", palette = get_palette("Paired",15), ylab = "Abundance") + 
  facet_wrap(~Group_Source, scales = "free_x", nrow = 3, ncol = 2)+
  labs(colour = "Class", fill = "Class") + 
  theme(axis.text.x = element_text(angle = 90, size = 5, hjust = 1, colour = "gray35"))

ARG_class_ab_group_facet
```

```{r}
setwd("~/OneDrive\ -\ Michigan\ State\ University/Zhang_lab/Mila_project/figures/smartchip")
ggsave(plot = ARG_class_ab_group_facet, "smartchip_class_relab_group_source.png", width = 20, height = 10)
```

#Aggregate phyloseq objects
##Aggregate samples by time points
```{r}
`%notin%` <- Negate(`%in%`)
physeq_day1 <- subset_samples(physeq, Time_tx%in%"Day -1")
physeq_week1 <- subset_samples(physeq, Time_tx%in%c("Week 1"))
physeq_week9 <- subset_samples(physeq, Time_tx%in%c("Week 9"))
```

Random forest FOR COMPONENTS
```{r}
physeq_component <- physeq_day1
# Make a dataframe of training data with OTUs as column and samples as rows
predictors <- t(otu_table(physeq_component)) 
dim(predictors)

# Make one column for our outcome/response variable 
response <- as.factor(sample_data(physeq_component)$Treatment)

# Combine them into 1 data frame
rf.data <- data.frame(response, predictors)

#Running random forest
library(randomForest)
set.seed(2)
classify.components <- randomForest(response~., data = rf.data, ntree = 5000, importance = T)
print(classify.components)
```

```{r fig.height=4}
# Make a data frame with predictor names and their importance
imp <- importance(classify.components)
imp <- data.frame(predictors = rownames(imp), imp)

# Order the predictor levels by importance
imp.sort_components <- arrange(imp, desc(MeanDecreaseAccuracy))
imp.sort_components$predictors <- factor(imp.sort_components$predictors, levels = imp.sort_components$predictors)

#Adding component number based on id table
library(purrr)
library(stringr)
imp.sort_components <- imp.sort_components%>% 
   mutate(Cluster_ID = map_chr(predictors, ~str_replace(.x, 
                                         pattern = "Cluster.", 
                                         replacement = "Cluster ")))

#imp.sort_components <- merge(imp.sort_components,taxonomy_df, by="Cluster_ID") %>%   select(Component,Control,Antibiotic,MeanDecreaseAccuracy,MeanDecreaseGini) %>% arrange(desc(MeanDecreaseAccuracy))

# Select the top 10 predictors
imp.20 <- imp.sort_components[1:20, ]

# ggplot
ggplot(imp.20, aes(x = reorder(predictors, -MeanDecreaseAccuracy), y = MeanDecreaseAccuracy)) + labs(x="Component")+
  geom_bar(stat = "identity", fill = "indianred") +
  coord_flip() +
  ggtitle("Most important metabolites for classifying fecal samples into IMM antibiotic treatment")
```
##Aggregate samples by time points
```{r}
`%notin%` <- Negate(`%in%`)
physeq_day1 <- subset_samples(physeq, Time_tx%in%"Day -1")
physeq_week1 <- subset_samples(physeq, Time_tx%in%c("Week 1"))
physeq_week9 <- subset_samples(physeq, Time_tx%in%c("Week 9"))
```

Random forest FOR COMPONENTS
```{r}
physeq_component <- physeq_week1
# Make a dataframe of training data with OTUs as column and samples as rows
predictors <- t(otu_table(physeq_component)) 
dim(predictors)

# Make one column for our outcome/response variable 
response <- as.factor(sample_data(physeq_component)$Treatment)

# Combine them into 1 data frame
rf.data <- data.frame(response, predictors)

#Running random forest
library(randomForest)
set.seed(2)
classify.components <- randomForest(response~., data = rf.data, ntree = 5000, importance = T)
print(classify.components)
```

```{r fig.height=4}
# Make a data frame with predictor names and their importance
imp <- importance(classify.components)
imp <- data.frame(predictors = rownames(imp), imp)

# Order the predictor levels by importance
imp.sort_components <- arrange(imp, desc(MeanDecreaseAccuracy))
imp.sort_components$predictors <- factor(imp.sort_components$predictors, levels = imp.sort_components$predictors)

#Adding component number based on id table
library(purrr)
library(stringr)
imp.sort_components <- imp.sort_components%>% 
   mutate(Cluster_ID = map_chr(predictors, ~str_replace(.x, 
                                         pattern = "Cluster.", 
                                         replacement = "Cluster ")))

#imp.sort_components <- merge(imp.sort_components,taxonomy_df, by="Cluster_ID") %>%   select(Component,Control,Antibiotic,MeanDecreaseAccuracy,MeanDecreaseGini) %>% arrange(desc(MeanDecreaseAccuracy))

# Select the top 10 predictors
imp.20 <- imp.sort_components[1:20, ]

# ggplot
ggplot(imp.20, aes(x = reorder(predictors, -MeanDecreaseAccuracy), y = MeanDecreaseAccuracy)) + labs(x="Component")+
  geom_bar(stat = "identity", fill = "indianred") +
  coord_flip() +
  ggtitle("Most important metabolites for classifying fecal samples into IMM antibiotic treatment")
```
##Aggregate samples by time points
```{r}
`%notin%` <- Negate(`%in%`)
physeq_day1 <- subset_samples(physeq, Time_tx%in%"Day -1")
physeq_week1 <- subset_samples(physeq, Time_tx%in%c("Week 1"))
physeq_week5 <- subset_samples(physeq, Time_tx%in%c("Week 5"))
physeq_week9 <- subset_samples(physeq, Time_tx%in%c("Week 9"))
```

Random forest FOR COMPONENTS
```{r}
physeq_component <- physeq_week5
# Make a dataframe of training data with OTUs as column and samples as rows
predictors <- t(otu_table(physeq_component)) 
dim(predictors)

# Make one column for our outcome/response variable 
response <- as.factor(sample_data(physeq_component)$Treatment)

# Combine them into 1 data frame
rf.data <- data.frame(response, predictors)

#Running random forest
library(randomForest)
set.seed(2)
classify.components <- randomForest(response~., data = rf.data, ntree = 5000, importance = T)
print(classify.components)
```

```{r fig.height=4}
# Make a data frame with predictor names and their importance
imp <- importance(classify.components)
imp <- data.frame(predictors = rownames(imp), imp)

# Order the predictor levels by importance
imp.sort_components <- arrange(imp, desc(MeanDecreaseAccuracy))
imp.sort_components$predictors <- factor(imp.sort_components$predictors, levels = imp.sort_components$predictors)

#Adding component number based on id table
library(purrr)
library(stringr)
imp.sort_components <- imp.sort_components%>% 
   mutate(Cluster_ID = map_chr(predictors, ~str_replace(.x, 
                                         pattern = "Cluster.", 
                                         replacement = "Cluster ")))

#imp.sort_components <- merge(imp.sort_components,taxonomy_df, by="Cluster_ID") %>%   select(Component,Control,Antibiotic,MeanDecreaseAccuracy,MeanDecreaseGini) %>% arrange(desc(MeanDecreaseAccuracy))

# Select the top 10 predictors
imp.20 <- imp.sort_components[1:20, ]

# ggplot
ggplot(imp.20, aes(x = reorder(predictors, -MeanDecreaseAccuracy), y = MeanDecreaseAccuracy)) + labs(x="Component")+
  geom_bar(stat = "identity", fill = "indianred") +
  coord_flip() +
  ggtitle("Most important metabolites for classifying fecal samples into IMM antibiotic treatment")
```
Random forest FOR COMPONENTS
```{r}
physeq_component <- physeq_week9
# Make a dataframe of training data with OTUs as column and samples as rows
predictors <- t(otu_table(physeq_component)) 
dim(predictors)

# Make one column for our outcome/response variable 
response <- as.factor(sample_data(physeq_component)$Treatment)

# Combine them into 1 data frame
rf.data <- data.frame(response, predictors)

#Running random forest
library(randomForest)
set.seed(2)
classify.components <- randomForest(response~., data = rf.data, ntree = 5000, importance = T)
print(classify.components)
```

```{r fig.height=4}
# Make a data frame with predictor names and their importance
imp <- importance(classify.components)
imp <- data.frame(predictors = rownames(imp), imp)

# Order the predictor levels by importance
imp.sort_components <- arrange(imp, desc(MeanDecreaseAccuracy))
imp.sort_components$predictors <- factor(imp.sort_components$predictors, levels = imp.sort_components$predictors)

#Adding component number based on id table
library(purrr)
library(stringr)
imp.sort_components <- imp.sort_components%>% 
   mutate(Cluster_ID = map_chr(predictors, ~str_replace(.x, 
                                         pattern = "Cluster.", 
                                         replacement = "Cluster ")))

#imp.sort_components <- merge(imp.sort_components,taxonomy_df, by="Cluster_ID") %>%   select(Component,Control,Antibiotic,MeanDecreaseAccuracy,MeanDecreaseGini) %>% arrange(desc(MeanDecreaseAccuracy))

# Select the top 10 predictors
imp.20 <- imp.sort_components[1:20, ]

# ggplot
ggplot(imp.20, aes(x = reorder(predictors, -MeanDecreaseAccuracy), y = MeanDecreaseAccuracy)) + labs(x="Component")+
  geom_bar(stat = "identity", fill = "indianred") +
  coord_flip() +
  ggtitle("Most important metabolites for classifying fecal samples into IMM antibiotic treatment")
```